(PE-11531) Allow upgrades from Puppet 4+

highb opened the pull request:
Prior to this commit we only allowed upgrades with this
module if the current puppet version was pre-4.
This commit removes that restriction and hopefully handles
all the edge cases involved in upgrade between new versions
and doesn't break FOSS. This feature should be tested
thoroughly.

highb commented:
I've tested this out with 3.8->2015.3.2 as well as 2015.2.1->2015.3.2 upgrades on centos7 and windows 2012 and it appears that `agent_upgrade.rb` passes. We'll definitely need to test more platforms.

jpartlow commented:
@highb tests are failing lint of manifests/osfamily/debian.pp - WARNING: variable not enclosed in {} on line 75.  I tried to push a commit fix to your branch but hit a permission issue (?)

Once that's fixed I'm seeing 70 spec errors around puppet_agent::service.  Are you not seeing those locally?

highb commented:
@jpartlow Hm, nope I don't think I was seeing those spec failures locally. Maybe something is off in my bundler environment, I'll poke around after I've had a change to check out these other PRs on the board.

highb commented:
@jpartlow OK, yeah was able to repro the spec failures locally. I'm not sure how I didn't see those before.  I'll get those fixed up.

jpartlow commented:
@highb Now we're seeing spec failures just for Puppet 3.7.5 and Puppet 4.0

highb commented:
@jpartlow Yep, I'm going to look into those once I've finished reviewing PRs.

MikaelSmith commented on manifests/init.pp:
```diff
@@ -66,16 +87,19 @@
     # On windows, our MSI handles the services
-    if $::osfamily != 'windows' {
+    # On PE AIO nodes, PE Agent nodegroup is managing the services
+    if $::osfamily != 'windows' and versioncmp($::clientversion, '4.0.0') < 0 {
```
Shouldn't this be restricted to is_pe as well then?

highb replied:
Oh yes, that'd be good to check.

thrnio commented on manifests/init.pp:
```diff
@@ -1,54 +1,75 @@
+    fail('upgrading requires at least Puppet 3.8')
   }
-  else {
+  elsif $is_pe and $::aio_agent_version != undef and $_package_version != undef and versioncmp("${::aio_agent_version}", "${_package_version}") >= 0 {
```
Why does this do nothing if up-to-date only if running on PE?

MikaelSmith replied:
We should probably still manage all the resources, and handle ensuring we don't make changes when the package installed is the correct version on a more granular level.

highb replied:
That makes sense to me. I think the only `is_pe` checks we should be doing are for the service check (since PE already manages the services) and for the PE only platforms (AIX, etc)

thrnio replied:
:+1:

MikaelSmith commented on manifests/init.pp:
```diff
@@ -66,16 +87,19 @@
+      package_version   => $package_version,
     }
 
     contain '::puppet_agent::prepare'
```
I think MODULES-3118 implies this needs to happen even if the aio_agent_version and package_version match.

highb commented:
@MikaelSmith Hm... we should probably think up a method for preventing that.

MikaelSmith commented:
Or support it. I'm ok with simply documenting it for now.

highb commented:
@MikaelSmith @thrnio I've finished up fixing the specs to test Puppet 4+, and addressed the lack of resource management in already upgraded PE nodes. I'm going to run some tests to verify PE doesn't break in some new and surprising way.

highb commented:
OK, one more spec fix is needed for pre-4 behavior.

MikaelSmith commented on Gemfile:
```diff
@@ -2,7 +2,7 @@ source ENV['GEM_SOURCE'] || 'https://rubygems.org'
 group :test do
   gem 'rake', '~> 10.4'
-  gem 'puppet', ENV['PUPPET_VERSION'] || '~> 3.8'
+  gem 'puppet', ENV['PUPPET_VERSION'] || '~> 4.0'
```
Why is this change needed?

highb replied:
Oh, it isn't really. That's just what I've been using for testing. All the tests should be specifying PUPPET_VERSION, though so it seems like a sensible default to me?

MikaelSmith replied:
Ok.

MikaelSmith commented on manifests/init.pp:
```diff
@@ -1,54 +1,72 @@
-# Upgrades Puppet 3.8 to Puppet 4+ (Puppet-Agent from Puppet Collection 1).
-# Makes the upgrade easier by migrating SSL certs and config files to the new
-# Puppet-Agent paths and removing deprecated settings that are no longer
+# Upgrades Puppet 3.8 and newer to the requested version.
```
Worth being clear that it will only upgrade to a puppet-agent package.

MikaelSmith commented on manifests/init.pp:
```diff
@@ -1,54 +1,72 @@
-    info('puppet_agent performs no actions on Puppet 4+')
+  if ($package_version == undef and versioncmp("${::clientversion}", '4.0.0') < 0) {
+    $_package_version = $puppet_agent::params::master_agent_version
+  } elsif ($package_version != undef and versioncmp("${::clientversion}", $package_version) != 0) {
```
I'm not clear what this check is doing?

Also, is there an attempt to skip installing if the correct version is already installed?

MikaelSmith replied:
Can we copy the `aio_agent_version` fact to this module, so OSS upgrades can use it? We should be checking whether `aio_agent_version` matches requested version, and not performing an install on platforms without package managers (or maybe just skip it on all platforms).

thrnio replied:
It's kind of silly IMO that the `aio_agent_version` fact is only shipped with PE.

MikaelSmith replied:
This may not be a major problem... rerunning the MSI on Windows takes about 5 seconds.

MikaelSmith replied:
Filed https://tickets.puppetlabs.com/browse/PA-241 to add it.

MikaelSmith replied:
Something like
```
diff --git a/manifests/install.pp b/manifests/install.pp
index 60cdde2..806f219 100644
--- a/manifests/install.pp
+++ b/manifests/install.pp
@@ -16,6 +16,7 @@ class puppet_agent::install(
 ) {
   assert_private()

+  if ($::aio_agent_version != $package_version) {
     if ($::operatingsystem == 'SLES' and $::operatingsystemmajrelease == '10') or ($::operatingsystem == 'AIX' and  $::architecture =~ /PowerPC_POWER[5,6,7]/) {
       contain puppet_agent::install::remove_packages

@@ -78,3 +79,4 @@ class puppet_agent::install(
       }
     }
   }
+}
```
would work. We could look at adding a custom fact, or just plan to address it for OSS in a future puppet-agent release.

highb replied:
Hm, I guess we should just set `_package_version` if `package_version` is set. The version mismatch check was from when I was thinking the module would do a no-op if an upgrade wasn't needed.

MikaelSmith replied:
This should only be needed for package providers that don't support versionable (Mac and Solaris 10), and Windows because we do an out-of-band upgrade. It looks like every provider we'll encounter that supports versionable also supports upgradeable, so no extra concerns there.

MikaelSmith replied:
I didn't see any changes corresponding to this problem (not repeating installs on Mac, Solaris 10, and Windows).

MikaelSmith commented on manifests/init.pp:
```diff
@@ -1,54 +1,72 @@
+      $_package_file_name = "${puppet_agent::package_name}-${_package_version}-1.osx${$::macosx_productversion_major}.dmg"
     } elsif $::osfamily == 'windows' {
       $_arch = $::kernelmajversion ?{
         /^5\.\d+/ => 'x86', # x64 is never allowed on windows 2003
```
This section needs to be updated to use the package version.

highb replied:
I recall going through a similar line of thought where I assumed that there would be a version on the package, but I think the Windows MSIs don't have versions specified on the package name?

MikaelSmith replied:
Might depend on where we get them. OSS versions do: https://downloads.puppetlabs.com/windows/

highb replied:
Interesting. Currently the windows prepare step is very PE-centric. I think it depends on pe_repo being setup on the master. We could just document that if you want to use this module with FOSS, you should re-name your MSI?

jpartlow replied:
There's something odd about that, because I thought windows upgrades were working in FOSS, and what we did for PE, was just deal with a) we needed to limit to installing windows packages provided with that pe version via pm.puppetlabs.com (downloaded locally to pe_repo), b) the fact that the master's cert was not trusted by windows boxes as far as transferring the windows package.

MikaelSmith replied:
Upgrades work, but they always get the latest release. With the change to specify a specific version, that logic needs to be updated.

highb replied:
Ah, right.

MikaelSmith commented on manifests/install.pp:
```diff
@@ -61,9 +64,16 @@
+    # Solaris 10/OSX package provider does not provide 'versionable'
+    # Package is removed above, then re-added as the new version here.
+    package { $::puppet_agent::package_name:
+      ensure => 'present',
```
How does 'present' ensure the correct package? Because we supply a source?

highb replied:
Yeah, since Solaris doesn't support versioned, we can't give a version here. So we just ensure that the source is the correct version.

highb commented:
@MikaelSmith Alright, I think I've addressed all your feedback and rebased the PR since it had some new merge conflicts.

MikaelSmith commented on manifests/init.pp:
```diff
@@ -1,51 +1,72 @@
-    info('puppet_agent performs no actions on Puppet 4+')
+  if ($package_version == undef and versioncmp("${::clientversion}", '4.0.0') < 0) {
+    $_package_version = $puppet_agent::params::master_agent_version
+  } else
```
Missing an open curly brace, caused CI to fail.

highb replied:
Odd, I remember fixing that.

MikaelSmith commented on manifests/install.pp:
```diff
@@ -7,9 +7,12 @@
 #
 class puppet_agent::install(
   $package_file_name = undef,
+  $package_version   = 'present',
```
Will this ever be used? We always pass an argument in init.pp, so seems like it will either be a version or undef.

highb commented:
I'll finish up the fixes for the specs and another rebase tomorrow.

highb commented:
Rebased again, now going to work on the platforms that need their packages removed before upgrading.

highb commented on manifests/init.pp:
```diff
@@ -1,46 +1,67 @@
+      $_package_file_name = "${puppet_agent::package_name}-${_package_version}-1.osx${$::macosx_productversion_major}.dmg"
     } elsif $::operatingsystem == 'aix' and $::architecture =~ /PowerPC_POWER[5,6,7]/ {
       $aix_ver_number = regsubst($::platform_tag,'aix-(\d+\.\d+)-power','\1')
       $_package_file_name = "${puppet_agent::package_name}-${puppet_agent::params::master_agent_version}-1.aix${aix_ver_number}.ppc.rpm"
```
Hm, this `${puppet_agent::params::master_agent_version}` should be `${package_version}`.

MikaelSmith replied:
Looks like it.

highb commented:
Specs were failing due to the linter not liking some double quoted strings w/o variables in them that I added. Fixed up. I've run the specs with Puppet 3.8 locally, hopefully nothing broke on 4.0

MikaelSmith commented on manifests/install/remove_packages.pp:
```diff
@@ -69,14 +67,22 @@
+    } elsif versioncmp("${::aio_agent_version}", "${::package_version}") < 0 {
+      if $::operatingsystem == 'Solaris' {
+        $packages = [
+          'PUPpuppet-agent',
```
This is necessary?

highb replied:
I think so, solaris is weird. I'll test it out to be sure.

Also, I still need to add SLES/AIX below to get around their weird, non-versionable package managers.

highb replied:
Oh no, this is going to cause a duplicate resource declaration...

MikaelSmith replied:
Why?

highb replied:
Ah, just ran a solaris upgrade and it *is* just `puppet-agent` now! Fixing...

MikaelSmith commented:
Looks good, except I think Windows will still unnecessarily re-install the MSI. You should guard windows/install.pp if aio_agent_version matches clientversion.

highb commented:
@MikaelSmith Then it will only be a problem on FOSS until we ship `aio_agent_version` with facter, right?

highb commented:
I'm testing all this out as I go along with pe_acceptance_tests/agent_upgrade.rb. I just added a minor change to agent_upgrade to add the package_version if we're upgrading from a PE_VER > 3.8.

highb commented:
Bah, didn't mean to push that up yet. Specs will fail because I haven't fixed some syntax issues and the specs still need another update. Update on testing: I've been working around the yum ensure => version issue that has been causing grief. It requires the `-1.el$majorversion` bit on the version, also just running 3.8.4/2015.3 tests takes quite a while...

highb commented:
Most recent changes have been tested on 3.8.4->2016.1 upgrade for centos7, solaris10 and sles10. Going to try out 2015.3.3->2016.1 with the same platforms plus OSX1010.

highb commented on .fixtures.yml:
```diff
@@ -1,6 +1,7 @@
 fixtures:
   repositories:
     stdlib: "https://github.com/puppetlabs/puppetlabs-stdlib.git"
+    transition: "https://github.com/puppetlabs/puppetlabs-transition.git"
```
Needed because otherwise we would have duplicate resource declarations on platforms that need to do the package remove/install dance to upgrade (Solaris/SLES/AIX)

MikaelSmith commented on manifests/install.pp:
```diff
@@ -50,20 +53,37 @@
-        package_file_name => $package_file_name,
-        source            => $::puppet_agent::source,
+    # Prevent re-running the batch install
+    if versioncmp("${::agent_version}", "${package_version}") < 0 {
```
So this blocks downgrades. That seems ok for now, since we don't officially support them.

MikaelSmith commented:
Failed CI. I'm :+1: on this in general though.

MikaelSmith commented on manifests/install/remove_packages.pp:
```diff
@@ -69,13 +73,24 @@
+      } else {
+        # We must use transition here because we would have a duplicate package
+        # declaration if we used a Package.
+        notify { "using puppetlabs-transition to remove ${old_package}: ${::operatingsystem} does not support versionable": }
```
You could use a name and message here to make the notify simpler.

MikaelSmith replied:
And don't you need to declare `package { $old_Package }` still?

highb replied:
Yeah, if I did declare it, that's where the duplicate resource errors come up.

highb replied:
Naming it would probably be a good call. :P

highb commented:
@MikaelSmith Yeah, because I've added a module the Puppetfile isn't updated, so I had to hack around in https://github.com/puppetlabs/pe_acceptance_tests/blob/2016.2.x/setup/agent_upgrade.rb#L112 to add the transition module. :/

highb commented:
Damn it, now the dev build version format is causing problems again. I guess I'll just make dev builds acceptable to the `package_version` regex.

MikaelSmith commented on manifests/init.pp:
```diff
@@ -1,49 +1,72 @@
   }
-  else {
+
+  if versioncmp("${::clientversion}", '3.8.0') < 0 {
```
This didn't rebase correctly.

MikaelSmith commented on spec/classes/puppet_agent_osfamily_aix_spec.rb:
```diff
@@ -1,7 +1,7 @@
 
 
-describe 'puppet_agent', :unless => Puppet.version < "3.8.0" || Puppet.version >= "4.0.0" do
+describe 'puppet_agent', :unless => Puppet.version < "3.8.0" do
```
We can stop guarding against version less than 3.8. Missed that in the commit that dropped the 3.8 guard.

MikaelSmith commented on spec/classes/puppet_agent_spec.rb:
```diff
@@ -29,31 +34,58 @@
-            it { is_expected.to contain_class('puppet_agent') }
-            it { is_expected.to contain_class('puppet_agent::params') }
-            if Puppet.version < "4.0.0"
+        if Puppet.version < "3.8.0"
```
This rebased incorrectly.

highb replied:
D'oh. Good catch.

MikaelSmith commented on manifests/install.pp:
```diff
@@ -50,20 +53,36 @@
+  } elsif ($::osfamily == 'RedHat') and ($package_version != 'present') {
+    # Workaround PUP-5802/PUP-5025
+    package { $::puppet_agent::package_name:
+      ensure => "${package_version}-1.el${::operatingsystemmajrelease}",
```
Yes, operatingsystemmajrelease is the right way to do this.

thrnio replied:
I thought 3.8 and stringified facts disabled was a requirement for this module?

MikaelSmith replied:
Stringified facts yes, but the 3.8 requirement (on agents) is going away.

highb commented:
Hopefully that's the last spec fix. Ever closer...

highb commented:
@MikaelSmith Commits squashed, I preserved the rest of the commit messages in the squash commit for posterity. Windows specs updated to check against `1.2.1.1`.
